javalib/src/main/scala/java/util/Arrays.scala

```diff
@@ -317,85 +418,179 @@ object Arrays {
-    if (a == null) 0
-    else a.foldLeft(1)((acc, x) => 31*acc + new java.lang.Double(x).hashCode)
+  def deepHashCode(a: Array[AnyRef]): Int = {
+    @inline
+    def getHash(elem: AnyRef): Int = {
+      elem match {
+        case elem: Array[AnyRef]  => deepHashCode(elem)
+        case elem: Array[Long]    => hashCode(elem)
+        case elem: Array[Int]     => hashCode(elem)
+        case elem: Array[Short]   => hashCode(elem)
+        case elem: Array[Char]    => hashCode(elem)
+        case elem: Array[Byte]    => hashCode(elem)
+        case elem: Array[Boolean] => hashCode(elem)
+        case elem: Array[Float]   => hashCode(elem)
+        case elem: Array[Double]  => hashCode(elem)
+        case _                    => if (elem == null) 0 else elem.hashCode
+      }
+    }
+    hashCodeImpl(a, getHash)
   }
 
-  def hashCode(a: Array[AnyRef]): Int = {
-    if (a == null) 0
-    else a.foldLeft(1)((acc, x) => 31*acc + (if (x == null) 0 else x.hashCode))
+  def deepEquals(a1: Array[AnyRef], a2: Array[AnyRef]): Boolean = {
+    @inline
+    def areDeepEquals(e1: AnyRef, e2: AnyRef): Boolean = {
+      if (e1 eq e2) true
+      else if (e1 == null || e2 == null) false
+      else {
+        (e1, e2) match {
+          case (a1: Array[AnyRef], a2: Array[AnyRef])   => deepEquals(a1, a2)
+          case (a1: Array[Long], a2: Array[Long])       => equals(a1, a2)
+          case (a1: Array[Int], a2: Array[Int])         => equals(a1, a2)
+          case (a1: Array[Short], a2: Array[Short])     => equals(a1, a2)
+          case (a1: Array[Byte], a2: Array[Byte])       => equals(a1, a2)
+          case (a1: Array[Char], a2: Array[Char])       => equals(a1, a2)
+          case (a1: Array[Boolean], a2: Array[Boolean]) => equals(a1, a2)
+          case (a1: Array[Float], a2: Array[Float])     => equals(a1, a2)
+          case (a1: Array[Double], a2: Array[Double])   => equals(a1, a2)
+          case (_, _)                                   => e1 == e2
+        }
+      }
+    }
+    if (a1 eq a2) true
+    else if (a1 == null || a2 == null || a1.length != a2.length) false
+    else a1.indices.forall(i => areDeepEquals(a1(i), a2(i)))
   }
 
-  def equals(a: Array[Boolean], b: Array[Boolean]): Boolean =
-    (a eq b) || (a != null && b != null && a.length == b.length &&
-        (0 until a.size).forall(i => a(i) == b(i)))
+  def toString(a: Array[Long]): String =
+    toStringImpl[Long](a, String.valueOf)
 
-  def equals(a: Array[Char], b: Array[Char]): Boolean =
-    (a eq b) || (a != null && b != null && a.length == b.length &&
-        (0 until a.size).forall(i => a(i) == b(i)))
+  def toString(a: Array[Int]): String =
+    toStringImpl[Int](a, String.valueOf)
 
-  def equals(a: Array[Byte], b: Array[Byte]): Boolean =
-    (a eq b) || (a != null && b != null && a.length == b.length &&
-        (0 until a.size).forall(i => a(i) == b(i)))
+  def toString(a: Array[Short]): String =
+    toStringImpl[Short](a, String.valueOf(_))
```

Comment:
I can't, `String.valueOf(x: Short)` is not defined. Putting the `(_)` make the necessary conversion to call `String.valueOf(x: Int)` instead.